File: 0110-balanced-binary-tree/0110-balanced-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        def height(root) -> int:
            if not root:
                return 0
            l = height(root.left)
            if l == float('inf'):
                return float('inf')
            r = height(root.right)
            if abs(l - r) > 1:
                return float('inf')
            return max(l, r) + 1
        if height(root) == float('inf'):
            return False
        return True

File: 0110-balanced-binary-tree/0110-balanced-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isBalanced(TreeNode* root) {
        if(checkBalance(root) == INT_MAX)
        {
            return false;
        }
        return true;
    }
private:
    int checkBalance(TreeNode* root)
    {
        if(!root)
        {
            return 0;
        }
        int l = checkBalance(root->left);
        if(l == INT_MAX)
        {
            return l;
        }
        int r = checkBalance(root->right);
        if(abs(l - r) > 1)
        {
            return INT_MAX;
        }
        return max(l, r) + 1;
    }
};